### PR TITLE
MINOR: Code cleanup Kafka Streams

### PR DESCRIPTION
**Pull request by sjhajharia**
## What
Code Cleanup in Streams Module

## Changes
Some common changes include

- Replace the Arrays.asList() with Collections.singletonList() wherever possible
- Cleaning up some if-else blocks
- Replacing some instances of String.builder() with String operations
- Replcing some format() with printf()

Thanks!

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorConvergenceTest.java [hunk 1]
```diff
@@ -623,11 +622,10 @@ private static void testForConvergence(final Harness harness,
...
         if (rebalancePending) {
-            final StringBuilder message =
-                new StringBuilder().append("Rebalances have not converged after iteration cutoff: ")
-                                   .append(iterationLimit)
-                                   .append(harness.history);
-            fail(message.toString());
+            final String message = "Rebalances have not converged after iteration cutoff: " +
+                    iterationLimit +
+                    harness.history;
+            fail(message);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorConvergenceTest.java]: If we move off `StringBuilder` than we can get rid of `String messsage`, too, and put the string directly into `fail(...)`

And we can change the formatting...

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorConvergenceTest.java [hunk 2]
```diff
@@ -569,11 +569,10 @@ private static void verifyBalancedAssignment(final Harness harness, final int sk
         final AssignmentTestUtils.TaskSkewReport taskSkewReport = AssignmentTestUtils.analyzeTaskAssignmentBalance(harness.clientStates, skewThreshold);
         if (taskSkewReport.totalSkewedTasks() > 0) {
             fail(
-                new StringBuilder().append("Expected a balanced task assignment, but was: ")
-                                   .append(taskSkewReport)
-                                   .append('\n')
-                                   .append(failureContext)
-                                   .toString()
+                    "Expected a balanced task assignment, but was: " +
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorConvergenceTest.java]: Seem we can use better formatting when removing `StringBuilder`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 3]
```diff
@@ -432,16 +429,15 @@ private static void validateAndAddActiveAssignments(final Set<TaskId> statefulTa
...
-                                       .append(activeTask)
-                                       .append(" on client ")
-                                       .append(entry)
-                                       .append(" but expected stateful tasks:")
-                                       .append(statefulTasks)
-                                       .append(" and stateless tasks:")
-                                       .append(statelessTasks)
-                                       .append(failureContext)
-                                       .toString()
+                        "Found an extra active task " +
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: Seem we can use better formatting when removing `StringBuilder`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 4]
```diff
@@ -396,27 +395,25 @@ private static void validateAndAddStandbyAssignments(final Set<TaskId> statefulT
...
                 throw new AssertionError(
-                    new StringBuilder().append("Found an extra standby task ")
-                                       .append(standbyTask)
-                                       .append(" on client ")
-                                       .append(entry)
-                                       .append(" but expected stateful tasks:")
-                                       .append(statefulTasks)
-                                       .append(failureContext)
-                                       .toString()
+                        "Found an extra standby task " +
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: Seem we can use better formatting when removing `StringBuilder`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 5]
```diff
@@ -396,27 +395,25 @@ private static void validateAndAddStandbyAssignments(final Set<TaskId> statefulT
...
                 throw new AssertionError(
-                    new StringBuilder().append("Found a standby task for stateless task ")
-                                       .append(standbyTask)
-                                       .append(" on client ")
-                                       .append(entry)
-                                       .append(" stateless tasks:")
-                                       .append(statelessTasks)
-                                       .append(failureContext)
-                                       .toString()
+                        "Found a standby task for stateless task " +
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: Seem we can use better formatting when removing `StringBuilder`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 6]
```diff
@@ -373,16 +373,15 @@ static void assertValidAssignment(final int numStandbyReplicas,
...
-                                   .append(numStandbyReplicas)
-                                   .append(" and max warmups ")
-                                   .append(maxWarmupReplicas)
-                                   .append(" standby replicas, stateful tasks:")
-                                   .append(statefulTasks)
-                                   .append(", and stateless tasks:")
-                                   .append(statelessTasks)
-                                   .append(failureContext)
-                                   .toString(),
+                    "Found some over- or under-assigned tasks in the final assignment with " +
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: Seem we can use better formatting when removing `StringBuilder`

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Couple of comments. Also, can you please rebase the PR to resolve merge conflict?

**Comment by mjsax** [comment]
@sjhajharia -- do you still have interest to finish this PR?

**Comment by sjhajharia** [comment]
Hey @mjsax 
Sorry, I was swamped a bit. Getting back to this, thanks for the review. I have updated the PR.
Could you pls help review again.
Thanks

File: streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyInnerJoinCustomPartitionerIntegrationTest.java [hunk 7]
```diff
@@ -125,8 +125,8 @@ public static void startCluster() throws IOException, InterruptedException {
             new KeyValue<>("ID123-4", "ID123-A4")
         );
 
-        final List<KeyValue<String, String>> table2 = asList(
-            new KeyValue<>("ID123", "BBB")
+        final List<KeyValue<String, String>> table2 = Collections.singletonList(
+                new KeyValue<>("ID123", "BBB")
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyInnerJoinCustomPartitionerIntegrationTest.java]: nit: avoid unnecessary reformatting

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 8]
```diff
@@ -374,19 +374,7 @@ static void assertValidAssignment(final int numStandbyReplicas,
...
-                                   .append(maxWarmupReplicas)
-                                   .append(" standby replicas, stateful tasks:")
-                                   .append(statefulTasks)
-                                   .append(", and stateless tasks:")
-                                   .append(statelessTasks)
-                                   .append(failureContext)
-                                   .toString(),
-                misassigned,
-                is(emptyMap()));
+            assertThat("Found some over- or under-assigned tasks in the final assignment with " + numStandbyReplicas + " and max warmups " + maxWarmupReplicas + " standby replicas, stateful tasks:" + statefulTasks + ", and stateless tasks:" + statelessTasks + failureContext, misassigned, is(emptyMap()));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]:
```suggestion
            assertThat(
                "Found some over- or under-assigned tasks in the final assignment with " + numStandbyReplicas +
                    " and max warmups " + maxWarmupReplicas + " standby replicas, stateful tasks:" + statefulTasks +
                    ", and stateless tasks:" + statelessTasks + failureContext, misassigned,
                is(emptyMap())
            );
```

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: Line too long.. Let's do something like this to make it easier to read.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java [hunk 9]
```diff
@@ -397,29 +385,11 @@ private static void validateAndAddStandbyAssignments(final Set<TaskId> statefulT
...
-                    new StringBuilder().append("Found a standby task for stateless task ")
-                                       .append(standbyTask)
-                                       .append(" on client ")
-                                       .append(entry)
-                                       .append(" stateless tasks:")
-                                       .append(statelessTasks)
-                                       .append(failureContext)
-                                       .toString()
-                );
+                throw new AssertionError("Found a standby task for stateless task " + standbyTask + " on client " + entry + " stateless tasks:" + statelessTasks + failureContext);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentTestUtils.java]: This is boarderline... maybe better to split into two lines? (similar below)

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java [hunk 10]
```diff
@@ -86,23 +86,23 @@ public static void main(final String[] args) throws IOException {
...
         streamsProperties.putAll(updated);
 
         if (!confirmCorrectConfigs(streamsProperties)) {
-            System.err.println(String.format("ERROR: Did not have all required configs expected  to contain %s, %s,  %s,  %s",
+            System.err.printf("ERROR: Did not have all required configs expected  to contain %s, %s,  %s,  %s%n",
                                              StreamsConfig.consumerPrefix(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG),
                                              StreamsConfig.producerPrefix(ProducerConfig.RETRIES_CONFIG),
                                              StreamsConfig.producerPrefix(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG),
-                                             StreamsConfig.producerPrefix(ProducerConfig.MAX_BLOCK_MS_CONFIG)));
+                                             StreamsConfig.producerPrefix(ProducerConfig.MAX_BLOCK_MS_CONFIG));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java]: Preferred formatting would be (but also ok to leave as-is):
```
            System.err.printf(
                "ERROR: Did not have all required configs expected  to contain %s, %s,  %s,  %s%n",
                StreamsConfig.consumerPrefix(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG),
                StreamsConfig.producerPrefix(ProducerConfig.RETRIES_CONFIG),
                StreamsConfig.producerPrefix(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG),
                StreamsConfig.producerPrefix(ProducerConfig.MAX_BLOCK_MS_CONFIG)
            );
```

**Review by mjsax — COMMENTED** [review]
Thanks for the PR update. Made another pass. Just nits, but given it's a cleanup PR, guess it ok to be nitty, as the whole PR is?

Btw: I though using `StringBuilder` is usually recommended as it more efficient? Does this not hold any longer? For this case, we are not on the critical code path so perf does not matter much, and I agree that `StringBuilder` make the code harder to read... Just curious.

**Review by mjsax — APPROVED** [review]

**Comment by mjsax** [comment]
Thanks for the PR. Merged to `trunk`.

**Pull request merged**